lib/resolvers/index.js

```diff
@@ -92,59 +91,85 @@ const filterResolvers = function (resolvers, excludes) {
     if (fieldResolvers instanceof GraphQLScalarType) {
-      wrapped[type] = fieldResolvers;
+      bound[type] = fieldResolvers;
       continue;
     }
 
-    if (!wrapped[type]) {
-      wrapped[type] = {};
+    if (!bound[type]) {
+      bound[type] = {};
     }
 
     for (const [field, resolverFunc] of Object.entries(fieldResolvers)) {
-      if (wrapped[type][field]) {
+      if (bound[type][field]) {
         continue;
       }
 
       if (['Query', 'Mutation'].indexOf(type) > -1) {
         debug(`memoized ${type}.${field}`);
-        wrapped[type][field] = memoize(type, field, resolverFunc.bind(bind));
+        bound[type][field] = memoize(type, field, resolverFunc.bind(bind));
       } else {
         // for types other than Query/Mutation - conditionally bind since in
         // some cases (Subscriptions and enum remaps) resolverFunc will be 
         // an object instead of a function
-        wrapped[type][field] = typeof resolverFunc === 'function' ? resolverFunc.bind(bind) : resolverFunc;
+        bound[type][field] = typeof resolverFunc === 'function' ? resolverFunc.bind(bind) : resolverFunc;
       }
     }
   }
-  return wrapped;
+  return bound;
 };
 
 /**
- * returns a proxy function that delegates execution to the input component's 
- * schema
- * @param {Object} component - the component to delegate execution to
- * @param {String} rootType - the root type whose field resolver will be 
- * executed 
- * @param {String} fieldName - the root type field whose resolver is proxied
- * @returns {Function} - a function whose signature is a normal GraphQL 
- * resolver function but whose implementation delegates execution to the 
- * component owning the resolver's implementation
+ * wraps non root type resolvers from the input resolver map that quick returns 
+ * if __typename is detected
+ * @param {object} resolvers - the input resolver map with which non root 
+ * resolvers will be wrapped
+ * @returns {object} the resulting resolver map with non root type resolvers 
+ * wrapped
  */
-const createProxyResolver = function (component, rootType, fieldName) {
-  const proxyResolver = async function (_, args, contextValue, info) {
-    debug(`delegating ${rootType}.${fieldName} to ${component.name}`);
-    return delegateToComponent(component, { contextValue, info });
-  };
+const wrapNonRootTypeResolvers = function (resolvers) {
+  const result = {};
 
-  proxyResolver.__isProxy = true;
+  for (const [type, fieldResolvers] of Object.entries(resolvers)) {
+    if (['Query', 'Mutation', 'Subscription'].indexOf(type) === -1) {
+      if (!result[type]) {
+        result[type] = {};
+      }
+      for (const [field, resolver] of Object.entries(fieldResolvers)) {
+        // handle non-root types that are enum remaps
+        if (typeof resolver !== 'function') {
+          result[type][field] = resolver;
+          continue;
+        }
+        if (field.startsWith('__')) {
```

Comment:
Not sure we need to define these separately...

IMO:

```js
result[type][field] = function (_, ...args) {
  if (_.__typename) {
    return _.__typename;
  }
  return resolver(_, ...args);
}
```

Reply:
I agree - except I thought it was clearer to separate them since the function signatures are technically different, and the return type is different when typename is detected:
* for __resolveType or __isTypeOf you return result.__typename
* for standard resolvers you return `result[info.fieldName]`

Reply:
Yeah, maybe the place to make the check is the return — to reproduce overall duplication and size of code.

Reply:
fixed